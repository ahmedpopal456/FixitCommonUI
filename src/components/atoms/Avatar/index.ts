import { AvatarBadge } from './Avatar';
export { AvatarBadge};