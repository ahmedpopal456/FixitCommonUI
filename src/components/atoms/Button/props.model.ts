import colors from '../../../theme/colors';

export interface Props {
    /**  To override default style */
    style?: any;
    /** Can be any color from the theme color palette */
    color?: keyof typeof colors;
    /** Callback function to be called when pressed */
    onPress(): void;
    /** Boolean value for outline button */
    $outline?: boolean;
    /** Boolean value for disabled button */
    disabled?: boolean;
    /** If true, the button will stretch to the screen width */
    block?: boolean;
    /** To pass custom icon */
    icon?: any;
    /** Icon Position */
    position?: "left" | "right";
    /** Button width */
    width?: number;
    /** Children which will be contained in the button (ie. Icon or string) */
    children: any;
    /** Shape of the ends of the button */
    shape?: "circle" | "square";
    /** Capitalizes the characters */
    $caps?: boolean;
    /** Used for testing */
    testID?: string;
  }