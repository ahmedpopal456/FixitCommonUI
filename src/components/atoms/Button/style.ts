import { Platform } from 'react-native';
import styled from "styled-components/native";
import custom from "styled-components";

import { Label } from "../Typography";
import { ButtonProps } from './button-props.model';
import colors from '../../../theme/colors';


/** Styled Components */
export const ButtonContainer = styled.TouchableOpacity`
  background-color: ${(props: ButtonProps) => getContainerColor(props)};
  margin: 5px;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  align-self: flex-start;
  padding: 7px;
  ${(props: ButtonProps) =>
        props.color != "transparent" &&
        `
      border-color: ${props.disabled ? colors.disabled : colors[props.color!]};
      }
      border-width: 1px;
      border-radius: ${props.shape === "square" ? '3px' : '100px'};
    `}
  ${(props: ButtonProps) =>
    props.width &&`
        width: ${props.width}px;
    `}
  ${(props: ButtonProps) =>
    Platform.OS === 'web' ?
        props.block && `
        flex: 1;
    `:
        props.block && `
        align-self: stretch;
    `}    
`;

export const IconWrapper = styled.View`
    padding: 8px;
`;

/** Helpers */
const getContainerColor = (props: ButtonProps) => {
    let color = colors[props.color!];

    if (props.$outline) {
        color = colors.transparent;
    } else if (props.disabled) {
        color = colors.disabled;
    }

    return color;
};

