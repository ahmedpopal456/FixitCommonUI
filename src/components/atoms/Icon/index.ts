import { Icon } from './Icon';
import { IconLibraries } from './IconLibraries';

export { Icon, IconLibraries };
