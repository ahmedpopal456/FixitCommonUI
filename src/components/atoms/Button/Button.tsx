import React from "react";

import { ButtonContainer, IconWrapper } from './style';
import { Label } from '../Typography'
import { ButtonProps } from "./button-props.model";

import colors from '../../../theme/colors';

/** 
 * Custom Fixit Button component
 * 
 * @param props - the button props
 * @returns a react native custom button component
 */
export const Button: React.FC<ButtonProps> = (props: ButtonProps): JSX.Element => {
  return (
    <ButtonContainer
      {...props}
    >
      {renderChildren(props)}
    </ButtonContainer>
  );
};

/**
 * 
 * @param props - The button's props
 * @returns The items that will be displayed in the button. ie. Label and/or Icon
 */
const renderChildren = (props: ButtonProps): JSX.Element => {
  return (
    <>
      {React.Children.map(props.children, (child: any) => {
        const childIsString: boolean =
          typeof child === "string" || child instanceof String;
        if (childIsString) {
          return <Label 
            color={getTextColor(props)}
            padding={6}
            fontSize={17}
            fontWeight={400}
            $caps={props.$caps}
            justification="center"
          >{child}</Label>;
        } else {
          const icon = React.cloneElement(child);
          return <IconWrapper>{icon}</IconWrapper>;
        }
      })}
    </>
  );
};

/**
 * @param props - The button's props
 * @returns a color value that the text should display
 */
const getTextColor = (props: ButtonProps): keyof typeof colors => {
  const bgColor = props.color!;
  let color: keyof typeof colors = bgColor == "primary" ? "accent" : "primary";

  if (props.$outline) {
      color = bgColor;
  } else if (props.disabled) {
      color = "light";
  }
  return color;
};

Button.defaultProps = {
  children: "",
  color: "primary",
  position: "left",
  shape: "square"
};