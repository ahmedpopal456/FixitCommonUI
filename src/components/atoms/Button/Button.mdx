---
name: Button
menu: 1 - Atoms
---

import { Props, Playground } from 'docz';
import { Button } from './Button.tsx';
import { Icon } from '../Icon/Icon.tsx';
import { View } from 'react-native';
import { DoczContainer } from '../../../utils/DoczContainer.tsx'

# Button
A Button is a clickable view to carry out actions such as clearing user input or submitting the form.

# Usage
```tsx
import { Button } from 'fixit-common-ui';

<Button onPress={doSomething}>
  Press Me
</Button>
```
## Props

<Props of={Button}></Props>

## Examples

<Playground>
  <DoczContainer>
    <Button onPress={() => void 0}>Press Me</Button>
    <Button onPress={() => void 0} color="accent">Press Me</Button>
    <Button onPress={() => void 0} $outline>Press Me</Button>
    <Button onPress={() => void 0} $outline color="accent">Press Me</Button>
  </DoczContainer>
</Playground>

## Disabled

<Playground>
  <DoczContainer>
    <Button onPress={() => void 0} disabled>
      Press Me
    </Button>
  </DoczContainer>
</Playground>

## Block

<Playground>
  <DoczContainer>
    <Button onPress={() => void 0} block>
      Press Me
    </Button>
  </DoczContainer>
</Playground>

## Width

<Playground>
  <DoczContainer>
    <Button onPress={() => void 0} width={150}>
      Press Me
    </Button>
    <Button onPress={() => void 0} width={200}>
      Press Me
    </Button>
    <Button onPress={() => void 0} width={300}>
      Press Me
    </Button>
  </DoczContainer>
</Playground>

## Caps

<Playground>
  <DoczContainer>
    <Button onPress={() => void 0} block $caps>capitalized</Button>
  </DoczContainer>
</Playground>


## Icon
<Playground>
  <DoczContainer>

    <Button onPress={() => void 0}>
      <Icon library="AntDesign" name="left" color="accent"/>
      Prev Page
    </Button>

    <Button onPress={() => void 0} color="primary">
      Next Page
      <Icon library="AntDesign" name="right" color="accent"/>
    </Button>

    <Button onPress={() => void 0} color="accent">
      <Icon library="AntDesign" name="barschart" color="primary"/>
    </Button>

    <Button onPress={() => void 0} color="green" shape="circle">
      <Icon library="AntDesign" name="plus" color="white"/>
    </Button>

    <Button onPress={() => void 0} color="transparent">
      <Icon library="AntDesign" name="heart" color="red"/>
    </Button>
    
  </DoczContainer>
</Playground>
