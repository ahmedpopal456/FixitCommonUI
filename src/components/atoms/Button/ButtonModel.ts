import { ReactElement } from 'react';
import { TextStyle } from 'react-native';
import colors from '../../../theme/colors';
import { IconProps } from '../Icon/IconModel';

export interface ButtonProps {
    /**  To override default style */
    style?: TextStyle & React.CSSProperties;
    /** Can be any color from the theme color palette */
    color?: keyof typeof colors;
    /** Callback function to be called when pressed */
    onPress(): void;
    /** Boolean value for outline button */
    outline?: boolean;
    /** Boolean value for disabled button */
    disabled?: boolean;
    /** If true, the button will stretch to the screen width */
    block?: boolean;
    /** Icon Position */
    position?: 'left' | 'right';
    /** Button width */
    width?: number;
    /** Children which will be contained in the button (ie. Icon or string) */
    children: string | ReactElement<IconProps>;
    /** Shape of the ends of the button */
    shape?: 'circle' | 'square';
    /** Capitalizes the characters */
    caps?: boolean;
    /** Used for testing */
    testID?: string;
    /** Padding */
    padding?: number;
  }
