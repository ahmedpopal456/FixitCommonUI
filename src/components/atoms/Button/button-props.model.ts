import { Icon } from '../Icon/Icon';
import colors from '../../../theme/colors';

export interface ButtonProps {
    /**  To override default style */
    style?: object;
    /** Can be any color from the theme color palette */
    color?: keyof typeof colors;
    /** Callback function to be called when pressed */
    onPress(): void;
    /** Boolean value for outline button */
    $outline?: boolean;
    /** Boolean value for disabled button */
    disabled?: boolean;
    /** If true, the button will stretch to the screen width */
    block?: boolean;
    /** To pass custom icon */
    icon?: typeof Icon;
    /** Icon Position */
    position?: "left" | "right";
    /** Button width */
    width?: number;
    /** Children which will be contained in the button (ie. Icon or string) */
    children: any;
    /** Shape of the ends of the button */
    shape?: "circle" | "square";
    /** Capitalizes the characters */
    $caps?: boolean;
    /** Used for testing */
    testID?: string;
  }