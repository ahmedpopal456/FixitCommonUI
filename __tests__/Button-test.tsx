import "react-native";
import React from "react";
import { Button } from "../src/components/atoms/Button";

// Note: test renderer must be required after react-native.
import renderer from "react-test-renderer";

it("renders default", () => {
  renderer.create(<Button onPress={() => void 0}>Button</Button>);
});

it("renders outline", () => {
  renderer.create(
    <Button onPress={() => void 0} outline>
      Button
    </Button>
  );
});

it("renders disabled", () => {
  renderer.create(
    <Button onPress={() => void 0} disabled>
      Button
    </Button>
  );
});

it("renders color", () => {
  renderer.create(
    <Button onPress={() => void 0} color="accent">
      Button
    </Button>
  );
});

it("renders width", () => {
  renderer.create(
    <Button onPress={() => void 0} width={30}>
      Button
    </Button>
  );
});

it("renders shape", () => {
  renderer.create(
    <Button onPress={() => void 0} shape="circle">
      Button
    </Button>
  );
});
it("matches snapshot", () => {
  const button = renderer
    .create(<Button onPress={() => void 0}>Button</Button>)
    .toJSON();
  expect(button).toMatchSnapshot();
});
