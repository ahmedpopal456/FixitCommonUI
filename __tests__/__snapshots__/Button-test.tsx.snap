// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches $outline snapshot 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "backgroundColor": "#00000000",
      "borderColor": "#33343E",
      "borderRadius": 3,
      "borderWidth": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "marginBottom": 5,
      "marginLeft": 5,
      "marginRight": 5,
      "marginTop": 5,
      "opacity": 1,
      "paddingBottom": 7,
      "paddingLeft": 7,
      "paddingRight": 7,
      "paddingTop": 7,
    }
  }
>
  <Text
    color="primary"
    style={
      Array [
        Object {
          "color": "#33343E",
          "fontSize": 17,
          "fontWeight": "600",
          "height": "100%",
          "paddingBottom": 8,
          "paddingLeft": 8,
          "paddingRight": 8,
          "paddingTop": 8,
        },
      ]
    }
  >
    Button
  </Text>
</View>
`;

exports[`matches color snapshot 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "backgroundColor": "#FFD14A",
      "borderColor": "#FFD14A",
      "borderRadius": 3,
      "borderWidth": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "marginBottom": 5,
      "marginLeft": 5,
      "marginRight": 5,
      "marginTop": 5,
      "opacity": 1,
      "paddingBottom": 7,
      "paddingLeft": 7,
      "paddingRight": 7,
      "paddingTop": 7,
    }
  }
>
  <Text
    color="accent"
    style={
      Array [
        Object {
          "color": "#33343E",
          "fontSize": 17,
          "fontWeight": "600",
          "height": "100%",
          "paddingBottom": 8,
          "paddingLeft": 8,
          "paddingRight": 8,
          "paddingTop": 8,
        },
      ]
    }
  >
    Button
  </Text>
</View>
`;

exports[`matches default snapshot 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "backgroundColor": "#33343E",
      "borderColor": "#33343E",
      "borderRadius": 3,
      "borderWidth": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "marginBottom": 5,
      "marginLeft": 5,
      "marginRight": 5,
      "marginTop": 5,
      "opacity": 1,
      "paddingBottom": 7,
      "paddingLeft": 7,
      "paddingRight": 7,
      "paddingTop": 7,
    }
  }
>
  <Text
    color="primary"
    style={
      Array [
        Object {
          "color": "#FFD14A",
          "fontSize": 17,
          "fontWeight": "600",
          "height": "100%",
          "paddingBottom": 8,
          "paddingLeft": 8,
          "paddingRight": 8,
          "paddingTop": 8,
        },
      ]
    }
  >
    Button
  </Text>
</View>
`;

exports[`matches disabled snapshot 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "backgroundColor": "#CECECE",
      "borderColor": "#CECECE",
      "borderRadius": 3,
      "borderWidth": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "marginBottom": 5,
      "marginLeft": 5,
      "marginRight": 5,
      "marginTop": 5,
      "opacity": 1,
      "paddingBottom": 7,
      "paddingLeft": 7,
      "paddingRight": 7,
      "paddingTop": 7,
    }
  }
>
  <Text
    color="primary"
    disabled={true}
    style={
      Array [
        Object {
          "color": "#EEEEEE",
          "fontSize": 17,
          "fontWeight": "600",
          "height": "100%",
          "paddingBottom": 8,
          "paddingLeft": 8,
          "paddingRight": 8,
          "paddingTop": 8,
        },
      ]
    }
  >
    Button
  </Text>
</View>
`;

exports[`matches shape snapshot 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "backgroundColor": "#33343E",
      "borderColor": "#33343E",
      "borderRadius": 100,
      "borderWidth": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "marginBottom": 5,
      "marginLeft": 5,
      "marginRight": 5,
      "marginTop": 5,
      "opacity": 1,
      "paddingBottom": 7,
      "paddingLeft": 7,
      "paddingRight": 7,
      "paddingTop": 7,
    }
  }
>
  <Text
    color="primary"
    style={
      Array [
        Object {
          "color": "#FFD14A",
          "fontSize": 17,
          "fontWeight": "600",
          "height": "100%",
          "paddingBottom": 8,
          "paddingLeft": 8,
          "paddingRight": 8,
          "paddingTop": 8,
        },
      ]
    }
  >
    Button
  </Text>
</View>
`;

exports[`matches width snapshot 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "backgroundColor": "#33343E",
      "borderColor": "#33343E",
      "borderRadius": 3,
      "borderWidth": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "marginBottom": 5,
      "marginLeft": 5,
      "marginRight": 5,
      "marginTop": 5,
      "opacity": 1,
      "paddingBottom": 7,
      "paddingLeft": 7,
      "paddingRight": 7,
      "paddingTop": 7,
      "width": 30,
    }
  }
>
  <Text
    color="primary"
    style={
      Array [
        Object {
          "color": "#FFD14A",
          "fontSize": 17,
          "fontWeight": "600",
          "height": "100%",
          "paddingBottom": 8,
          "paddingLeft": 8,
          "paddingRight": 8,
          "paddingTop": 8,
        },
      ]
    }
  >
    Button
  </Text>
</View>
`;
